metomi/rose/popen.py

```diff
@@ -128,11 +128,41 @@ class RosePopener:
     }
     ENVS_OF_CMDS = {"editor": ["VISUAL", "EDITOR"]}
 
-    @classmethod
-    def list_to_shell_str(cls, args):
-        if not args:
-            return ""
-        return " ".join([re.sub(r"([\"'\s])", r"\\\1", arg) for arg in args])
+    @staticmethod
+    def shlex_join(args: Iterable[str]) -> str:
+        """Basically a back-port of shlex.join(), needed for py 3.7
+
+        Examples:
+        >>> RosePopener.shlex_join([])
+        ''
+        >>> RosePopener.shlex_join(["echo", "-n", "Multiple words"])
+        "echo -n 'Multiple words'"
+        >>> RosePopener.shlex_join(["ls", "my_dir;foiled_injection"])
+        "ls 'my_dir;foiled_injection'"
+        >>> RosePopener.shlex_join(["what", "about", "globs*"])
+        "what about 'globs*'"
+        """
+        return " ".join(shlex.quote(arg) for arg in args)
+
+    @staticmethod
+    def _list_to_shell_str(args: Iterable[str]) -> str:
+        """NOTE: This is not very safe. Keeping for back compat, but ideally
+        use self.shlex_join() instead going forward.
```

Comment:
I don't think we can get rid of this, Rose allows users to specify commands to be run directly in the shell e.g:

```ini
[command]
default=echo * | sed 's/.../.../' || true
```

These commands can use all shell features. Cylc gets around this by running jobs in a Bash script but the end result is exactly the same. Under the circumstances this approach is perfectly "safe", Rose is essentially entering the command you told it to run directly into the shell as if you had done so yourself.

So I would update this docstring because this isn't something we can phase out, should make it clear when this method is needed and when we can avoid it.

E.G:

This permits shell features such as chained commands and globs required which are required for running user-defined code. Where this is not needed use `self.shlex_join` in preference.

Reply:
Ok, that makes sense, will add to docstring so there is no ambiguity about this in future

Reply:
https://github.com/metomi/rose/compare/f7cc78df733d6a927b0f6eda61c71ccf5e9f4041..d2fae0621f730cd4dd926628d14da4c5b7ac8dbe